pkg/disttask/framework/scheduler/balancer_test.go

```diff
@@ -0,0 +1,433 @@
+				{ID: 4, ExecID: "tidb2", Concurrency: 8, State: proto.SubtaskStateRunning},
+				{ID: 5, ExecID: "tidb3", Concurrency: 8, State: proto.SubtaskStatePending},
+			},
+		},
+		// task 3 require balance, but no eligible node, so it's not balanced, and
+		// used slots are not updated
+		// used slots will be {tidb1: 16, tidb2: 16, tidb3: 8}
+		{
+			subtasks: []*proto.Subtask{
+				{ID: 6, ExecID: "tidb4", Concurrency: 16, State: proto.SubtaskStatePending},
+				{ID: 7, ExecID: "tidb4", Concurrency: 16, State: proto.SubtaskStatePending},
+			},
+			expectedSubtasks: []*proto.Subtask{
+				{ID: 6, ExecID: "tidb4", Concurrency: 16, State: proto.SubtaskStatePending},
+				{ID: 7, ExecID: "tidb4", Concurrency: 16, State: proto.SubtaskStatePending},
+			},
+		},
+		// task 4 require balance
+		// used slots will be {tidb1: 16, tidb2: 16, tidb3: 16}
+		{
+			subtasks: []*proto.Subtask{
+				{ID: 8, ExecID: "tidb1", Concurrency: 8, State: proto.SubtaskStatePending},
+			},
+			expectedSubtasks: []*proto.Subtask{
+				{ID: 8, ExecID: "tidb3", Concurrency: 8, State: proto.SubtaskStatePending},
+			},
+		},
+	}
+	ctx := context.Background()
+
+	manager, err := NewManager(ctx, mockTaskMgr, "1")
+	require.NoError(t, err)
+	manager.slotMgr.updateCapacity(16)
+	manager.nodeMgr.managedNodes.Store(&[]string{"tidb1", "tidb2", "tidb3"})
+	b := newBalancer(Param{
+		taskMgr: manager.taskMgr,
+		nodeMgr: manager.nodeMgr,
+		slotMgr: manager.slotMgr,
+	})
+	for i := range taskCases {
+		taskID := int64(i + 1)
+		sch := mock.NewMockScheduler(ctrl)
+		sch.EXPECT().GetTask().Return(&proto.Task{ID: taskID}).AnyTimes()
+		manager.addScheduler(taskID, sch)
+	}
+	require.Len(t, manager.getSchedulers(), 4)
+
+	// fail fast if balance failed on some task
+	manager.getSchedulers()[0].(*mock.MockScheduler).EXPECT().
+		GetEligibleInstances(gomock.Any(), gomock.Any()).Return(nil, errors.New("mock error"))
+	b.balance(ctx, manager)
+	require.True(t, ctrl.Satisfied())
+
+	// balance multiple tasks
+	for i, c := range taskCases {
+		taskID := int64(i + 1)
+		if !assert.ObjectsAreEqual(c.subtasks, c.expectedSubtasks) {
+			gomock.InOrder(
+				manager.getSchedulers()[i].(*mock.MockScheduler).EXPECT().
+					GetEligibleInstances(gomock.Any(), gomock.Any()).Return(nil, nil),
+				mockTaskMgr.EXPECT().GetActiveSubtasks(gomock.Any(), taskID).Return(c.subtasks, nil),
+				mockTaskMgr.EXPECT().UpdateSubtasksExecIDs(gomock.Any(), gomock.Any()).Return(nil),
+			)
+		} else {
+			gomock.InOrder(
+				manager.getSchedulers()[i].(*mock.MockScheduler).EXPECT().
+					GetEligibleInstances(gomock.Any(), gomock.Any()).Return(nil, nil),
+				mockTaskMgr.EXPECT().GetActiveSubtasks(gomock.Any(), taskID).Return(c.subtasks, nil),
+			)
+		}
+	}
+	b.balance(ctx, manager)
+	require.Equal(t, map[string]int{"tidb1": 16, "tidb2": 16, "tidb3": 16}, b.currUsedSlots)
+	require.True(t, ctrl.Satisfied())
+	for _, c := range taskCases {
+		require.Equal(t, c.expectedSubtasks, c.subtasks)
+	}
```

Comment:
It is better to add a comment😂
```suggestion
	for _, c := range taskCases {
		// c.subtasks could be changed in b.balance
		require.Equal(t, c.expectedSubtasks, c.subtasks)
	}
```

Reply:
see https://github.com/pingcap/tidb/pull/49758/files#diff-25ae0070584471094d3aa22362f14ac0d9f2f65f22709dd2ddf3902153d4c0dfR251